sui_programmability/examples/nfts/sources/random_word.move

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// A NFT shop that creates NFTs with random words and can be minted without going through consensus.
+/// It is an example of a secure auto generated NFTs factory, e.g. for selling game items.
+///
+/// The NFT creator defines the possible words and their weights.
+/// Anyone can purchase a word from the shop, chosen randomly for the user according to the distribution defined by
+/// the weights.
+///
+module nfts::random_word {
+    use sui::object::{Self, UID, ID};
+    use sui::tx_context::{Self, TxContext};
+    use std::string::String;
+    use sui::coin::Coin;
+    use sui::sui::SUI;
+    use sui::coin;
+    use sui::randomness;
+    use std::option;
+    use std::vector;
+
+    const EWrongShop: u64 = 1;
+    const EWrongRandomness: u64 = 2;
+    const EInvalidRndLength: u64 = 3;
+    const EInvalidSize: u64 = 4;
+
+    /// Instances of this type will be immutable objects, thus users could use them without going through consensus.
+    struct Shop has key {
+        id: UID,
+        issuer: address,
+        words: vector<String>,
+        weights: vector<u8>,
+        total_weight: u64,
+    }
+
+    struct Ticket has key {
+        id: UID,
+        shop_id: ID,
+        randomness_id: ID,
+    }
+
+    /// The "word NFT" which links to the creating shop and has the random word chosen for the user.
+    struct Word has key {
+        id: UID,
+        word: String,
+        creator: ID,
+    }
+
+    struct RANDOMNESS_WITNESS has drop {}
+
+    /// Create a new word NFTs shop for a given set of words and their weights.
+    public entry fun create(words: vector<String>, weights: vector<u8>, ctx: &mut TxContext) {
+        assert!(vector::length(&words) == vector::length(&weights), EInvalidSize);
+        let total_weight: u64 = 0;
+        let i = 0;
+        while (i < vector::length(&weights)) {
```

Comment:
Nit: I believe it's common to get length once and compare against a const value. Not sure how much it affects gas.